### PR TITLE
Feature/issue 977 some variadic templating

### PR DESCRIPTION
**Pull request by dpsimpson**
## Summary

I replace the code in `return_type`, `partials_return_type` and `include_summand` to use variadic template parameters. This should make it unnecessary to keep expanding the code every time we need a function that takes more arguments.

I'm not 100% happy with the base case in `include_summand`, I feel like it should be `include_summand<bool propto>` but I couldn't make partial specialization work with a non-type template parameter. I imagine that it's something dumb I did.

The two return_type programs promote everything lower in precedence to `double`, which is consistent with the existing behaviour. 

## Tests

I wrote new tests for `partials_return_type` and `include_summand`.  I expanded the test for `return_type`.
- All the tests have multiple types (including `var`)
- All the `*return_type` tests test `int` and `float` (consistent with existing tests) and put `var`s in various places in the template arguments.
- The `include_summand` tests cover no types + `propto=false`, `propto=true` + `double` types, `propto = true` + a mixture of `double` and `var`

This is a kinda important piece of code, so let me know if I missed any tests

## Side Effects

None. 

## Checklist

- [x] Math issue #977 

- [ ] Copyright holder: Daniel Simpson 

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`) (Passes)
    - docs build, (`make doxygen`) (Error)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`) (Passes)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ]the new changes are tested

### CODE REVIEW

File: stan/math/prim/scal/meta/partials_return_type.hpp [hunk 1]
```diff
@@ -7,16 +7,19 @@
 
 namespace stan {
```
Comment (syclik) [inline stan/math/prim/scal/meta/partials_return_type.hpp]: Needs documentation.

@dpsimpson, I'm not attributing the lack of doc to you. This is just some of the technical debt we've accumulated over the years. I'm also not forcing you to write doc. If you're happy to, then please, by all means. If you don't want to, no worries; we can find a volunteer to write this doc.

File: stan/math/prim/scal/meta/return_type.hpp [hunk 2]
```diff
@@ -9,15 +9,20 @@ namespace stan {
 /**
  * Metaprogram to calculate the base scalar return type resulting
  * from promoting all the scalar types of the template parameters.
+ * Note: Sub-double types are promoted to double.
  */
-template <typename T1, typename T2 = double, typename T3 = double,
-          typename T4 = double, typename T5 = double, typename T6 = double>
+
```
Comment (syclik) [inline stan/math/prim/scal/meta/return_type.hpp]: remove this blank line.

File: stan/math/prim/scal/meta/return_type.hpp [hunk 3]
```diff
@@ -9,15 +9,20 @@ namespace stan {
 /**
  * Metaprogram to calculate the base scalar return type resulting
  * from promoting all the scalar types of the template parameters.
+ * Note: Sub-double types are promoted to double.
```
Comment (syclik) [inline stan/math/prim/scal/meta/return_type.hpp]: what is "sub-double"?

After rereading this a few times, is this what you're saying?
"Note: C++ primitive types are promoted to double."

**Comment by syclik** [comment]
I think my PR review disappears when my it's dismissed or approved, so I wanted to reiterate here.

Thank you! This is awesome.

Re: your question about tests. There's enough coverage. Thank you for writing the PR clearly and asking about testing. When I read the PR, I wanted to make sure we tested the meta programs in the unit tests with enough template parameters. The reason is that even though it's valid and legal, we may end up with practical limitations in compilers. It'd be great to know that we just couldn't instantiate this function rather than something downstream and we're trying to figure out why it won't work. What I was going to ask for was to include a test that had as many template parameters as you need for new work, but you already included the test!

File: stan/math/prim/scal/meta/return_type.hpp (hunk 3 above)
Comment (dpsimpson) [reply stan/math/prim/scal/meta/return_type.hpp]: That’s probably what I’m saying. Anything that `promote_args` will promote to double will be. I didn’t check if that’s all the primitive types.

File: stan/math/prim/scal/meta/include_summand.hpp [hunk 4]
```diff
@@ -26,28 +26,25 @@ namespace math {
  * proportionality constant.
  * @tparam T1 First
```
[the remark this answers is not in the document]
Comment (syclik) [reply stan/math/prim/scal/meta/include_summand.hpp]: THANK YOU! (You've had enough correspondence with me; shouting is rare. This deserves shouting.)

File: stan/math/prim/scal/meta/include_summand.hpp [hunk 5]
```diff
@@ -26,28 +26,25 @@ namespace math {
...
-          typename T9 = double, typename T10 = double>
+template <bool propto, typename T = double, typename... T_pack>
 struct include_summand {
+  enum {
+    value = (!stan::is_constant<typename scalar_type<T>::type>::value
+             || include_summand<propto, T_pack...>::value)
+  };
+};
+
+// T defaults to double (specializations inherrit defaults)
```
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply stan/math/prim/scal/meta/include_summand.hpp]: Would this work?

```
template <bool propto>
bool include_summand<propto, double> {
  enum { value = false; }
};
```

The compiler will first find the best match among the base templates, then when it has the base template match, will find the best match among specializations.

[the remark this answers is not in the document]
Comment (bob-carpenter) [reply stan/math/prim/scal/meta/include_summand.hpp]: Ack, no.  Never mind  That won't work for any instantiation other than `double`.  What was I thinking?  The resolution rule is right, but I wasn't applying it correctly.  This is a puzzle.  I bet it has a known answer.

File: stan/math/prim/scal/meta/partials_return_type.hpp (hunk 1 above)
Comment (dpsimpson) [reply stan/math/prim/scal/meta/partials_return_type.hpp]: Documented. Which is good because I actually misunderstood what the return type was and made the unit test incorrectly. I now check that it sends `var` -> `double` and `fvar<var>` -> `double`.

Do I need to test that it sends eg `fvar<float>` -> `double`?  I think this is implied by the other tests.

It would be good to add "when this should be used" to the doc, because it's mainly popping up in modern code.

I'm also not 100% sure that the use of it is always safe for higher-order partials. For example, I'm not sure what lines 93-98 of @VMatthijs `normal_id_glm_lpdf` code. https://github.com/VMatthijs/math/blob/f9d380d7a65ba4978753164c4bf11dcd08dc6f7e/stan/math/prim/mat/prob/normal_id_glm_lpdf.hpp

In that case, if there is a `fvar<var>` in the template parameters, those lines will construct a column vector of `var`s, which are then assigned like doubles. Will this work?  If not, I'll make an issue.

**Comment by dpsimpson** [comment]
I've made all of the changes requested.  
- There's something going wrong with the doxygen in `partials_return_type` that I can't work out how to fix.
- I made a more paranoid test for `return_type` to prove to myself that this will compile easily with 21 template parameters, which is more than anyone is likely to need.
- Doc for `partials_return_type` involved me learning what it does, which wasn't what I expected. There was an incorrect unit test that I fixed. I also made sure it had the correct behaviour for `fvar<var>` because that's the only case where it's not a `double`.
- I added another test to `include_summand` just in case I messed up the default template parameter (I didn't).

Apologies for the churn. I kept forgetting to pull after Jenkins did it's reformatting.

Tests are running as we speak, and I need to double check that I have the copyright and not the Toronto Trustees, which will be done next week.

**Comment by bob-carpenter** [comment]
The targets of interest are:

`int, double, var, fvar<double>, fvar<fvar<double>>, fvar<var>, fvar<fvar<var>>`

The only thing we actually use now in the langauge are the first three.

**Comment by bob-carpenter** [comment]
Oh, and no mixtures.  Every function gets arguments that are `int`, `double`, and at most one autodiff type.  

None of the autodiff promotions or demotions make sense in context, and you can show inductively they never arise from the Stan language.

File: test/unit/math/prim/scal/meta/return_type_test.cpp [hunk 6]
```diff
@@ -15,3 +17,26 @@ TEST(MetaTraits, ReturnTypeFloat) {
 TEST(MetaTraits, ReturnTypeInt) {
   test::expect_same_type<double, return_type<int>::type>();
 }
+
+TEST(MetaTraits, ReturnTypeDoubleMany) {
+<<<<<<< HEAD
```
Comment (syclik) [inline test/unit/math/prim/scal/meta/return_type_test.cpp]: Looks like some merge didn't go as expected?

**Comment by syclik** [comment]
I was thinking about the testing. We could reduce the testing a bit. This doesn't have to be done for this PR, but here's what I'm thinking:

- since the implementation is now recursive using variadic templates, we could get away with testing the base case, then a test to make sure the 3 argument function passes through. 
- since we've run into compiler implementation errors, to have confidence, I'd be happy with one test (per autodiff type) that is has as many arguments that can be generated from Stan. That won't be perfect, but it should tell us if the template depth doesn't get implemented correctly on one of our compilers.

**Comment by bbbales2** [comment]
@dpsimpson Is "I'm not sure what lines 93-98 of @VMatthijs normal_id_glm_lpdf code." still a problem?

I only glanced at it but I wasn't sure what should be happening for higher order autodiff. I reviewed that so it's entirely possible I goofed. Lemme know if you still think it's shaky and I'll investigate. @VMatthijs should be aware too.

**Comment by dpsimpson** [comment]
@bbbales2 I honestly don't know. The simplest version of lines 82-90 of [`normal_id_glm_lpdf.hpp`](https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/prob/normal_id_glm_lpdf.hpp)
occurs when `sigma` has length one. In this case the `sigma` part of the code chunk is equivalent to
```
Array<T_partials_return, 1, 1> = value_of(sigma]);
```
`T_partials_return` is either `double`, in which case the code is fine. Or it's a forward autodiff type, in which case I'm not sure if that assignment makes sense. 

tl;dr: I don't know if there's a problem, but @VMatthijs' naming makes it pretty clear they're being thought of as doubles, so I thought I'd check with someone!  If no one knows, I can make an issue.

**Comment by dpsimpson** [comment]
@syclik I am doing the comprehensive tests, because I'm big into "best practice" right now :-p The test now check each of the seven types @bob-carpenter listed individually and inside a group of 10 template parameters (which is slightly larger than the current max number in the Stan code).

One odd behaviour involves the following two tests for `return_type` (five parameter simplifications of the actual tests)
```
TEST(MetaTraits, ReturnTypeFvarVarFiveParams) {
  test::expect_same_type<
      fvar<var>, return_type<double, fvar<var>,
                       float, fvar<var>, int>::type>();
}

TEST(MetaTraits, ReturnTypeFvarFvarVarFiveParams) {
  test::expect_same_type<
      fvar<fvar<var> >, return_type<double, fvar<fvar<var> >, 
                       float, float, fvar<fvar<var> >, int>::type>();
}
```
These tests only pass if I include `stan/math/rev/scal.hpp`. This is true for both the old implementation and the new implementation. (Without the inclusion the new implementation fails in all cases, while the old implementation will pass if the autodiff variable is the first type passed.)

Everything I know about the problem is below. I'm a little worried because I don't understand the problem, but the tests pass with the file included.

(tl;dr: Boost is doing something weird, which is magically fixed with the include, but not fixed if I only include the functions that work in the `boost::math` namespace.)

_I am very happy to not worry about this any further, but I just thought I'd flag it in case someone thinks it's indicative of a wider problem. (Yes - I'm aware that for 40 lines of code, this is a very high maintenance PR. Sorry!)_

**Detailed explanation of what happens**

Without that inclusion the following occurs
```
typedef typename boost::math::tools::promote_args<double, fvar<var> >::type type; // type is double
typedef typename boost::math::tools::promote_args<fvar<var>, double >::type type; // type is fvar<var>
```
If `stan/math/rev/scal.hpp` is included, the behaviour is symmetric (which needs to be true to pass these two tests) and the type is `fvar<var>` as expected.  

(The same behaviour occurs for `fvar<fvar<var> >`.)

The thing that slightly concerns me is that this has to be a boost problem. The bit where it fails is Line 104 of [`promotion.hpp`](https://github.com/stan-dev/math/blob/develop/lib/boost_1.66.0/boost/math/tools/promotion.hpp)
which covers the case where at least one of the two types being promoted is user defined. If the second type is not floating point AND the first type can be converted to the second, then the second type is selected. Otherwise it's the first type.

So this means that something in `stan/math/rev/scal.hpp` is making boost think a double is promotable to a `fvar<var>` or `fvar<fvar<var> >`. It is not one of the files `stan/math/rev/scal/fun/boost_*` because I included them and it didn't fix the problem. 

I can't make heads nor tails of boost's `boost/type_traits/is_convertible.hpp`  function, so I'm not sure why this is happening.

**Comment by dpsimpson** [comment]
I am completely at a loss to work out what this Doxygen warning means:

```partials_return_type.hpp:21, GNU C Compiler 4 (gcc), Priority: High
--
explicit link request to 'type' could not be resolved (warning treated as error, aborting now)
```

The referenced line is 

> ` * containers of <code>var</code>s.`

Any suggestions?

**Comment by bob-carpenter** [comment]
> On Aug 13, 2018, at 12:15 AM, dpsimpson <notifications@github.com> wrote:
> 
> I am completely at a loss to work out what this Doxygen warning means:
> 
> --
> explicit link request to 'type' could not be resolved (warning treated as error, aborting now)
> 
> The referenced line is
> 
> * containers of <code>var</code>s.
> 
> Any suggestions?

Drop the markup and hope for the best.  You should be able to run locally.

**Comment by bob-carpenter** [comment]
You cut and paste the answer:

  If the second type is not floating point AND the first
  type can be converted to the second, then the second type
  is selected. Otherwise it's the first type.

> typedef typename boost::math::tools::promote_args<double, fvar<var> >::type type; // type is double

Without the declaration of the constructor var(double), the compiler won't be able to tell that double is promotable to fvar<var>, so the type is double according to the above rule.

> typedef typename boost::math::tools::promote_args<fvar<var>, double >::type type; // type is fvar<var>

Again, because it's not assignable, the first gets returned.

Then, when you include rev/scal.hpp, you get the declaration var(double) and the assignments work and it's all copacetic.

**Comment by dpsimpson** [comment]
@bob-carpenter The Doxygen problem was having nested template parameters, so `fvar<T>` was fine but `partials_return_type<fvar<T> >` threw a warning/error.

I still don't understand why double was convertible to `var` with `math/rev/core.hpp` inlcuded, but double was only convertible to `fvar<var>` if `math/rev/scal.hpp` was included. But I can live with not knowing.

@syclik Thanks for nudging me into all of those tests! It turns out the original was missing some `#include`s which meant `partials_type<>` wasn't properly specialized for `fvar<T>`.  The code now has the correct behaviour `partials_return_type<fvar<T> >::type` has type `T`.

In the previous implementation (without the variadic templates) this was also incorrect, although I don't think it's been used so it's probably ok.

Bug was found by writing extensive tests that passed and then adding an `include<stan/math.hpp>` to each file and discovering that the tests no longer passed!

**Comment by dpsimpson** [comment]
Ok. I have an answer to why I needed that `math/rev/scal.hpp` include. The declaration for `fvar` is in `math/fwd/core/fvar.hpp` and the constructor  `fvar(const T& v)` calls `is_nan(v)`. 

Now that file includes `stan/math/prim/scal/fun/is_nan.hpp` but doesn’t include either the `rev` or `fwd` equivalents. This seems like an oversight. But I’ll leave it to people more knowledgeable to decide if it should change.

**Comment by syclik** [comment]
@dpsimpson, thank you for going through this in so much depth. I think you're right about the `is_nan.hpp` include in `fvar.hpp`, but I think the correct fix is to remove the `is_nan` check from the constructor. I introduced that, but as a debug measure when trying to determine if we're constructing `fvar`s properly on all of our operations. That shouldn't be there. I just created a new issue to fix that separately: #983.

Thanks for sticking with it and helping us pay down some of our technical debt.

File: stan/math/prim/scal/meta/partials_return_type.hpp (hunk 1 above)
Comment (syclik) [reply stan/math/prim/scal/meta/partials_return_type.hpp]: Hmm... is that right? I thought `partials_return_type` was the type of the partial itself.

For a `stan::math::var`, the partial should be `double`.

For a `stan::math::fvar<double>`, it should be `double`. (I think if it was `fvar<float>`, it should return `double`, but we don't need to check this at every level of nesting.)

For a `stan::math::fvar<var>`, isn't it `var`? Or am I missing something? The directional derivative returned is a `var` (so we can autodiff through that).

File: stan/math/prim/scal/meta/partials_return_type.hpp [hunk 7]
```diff
@@ -2,21 +2,50 @@
...
+#include <stan/math/fwd/scal/meta/partials_type.hpp>
 #include <stan/math/prim/scal/meta/scalar_type.hpp>
 #include <boost/math/tools/promotion.hpp>
 
 namespace stan {
 
-template <typename T1, typename T2 = double, typename T3 = double,
-          typename T4 = double, typename T5 = double, typename T6 = double>
+/**
+ * Template metaprogram to calculate the base scalar return type resulting
```
Comment (syclik) [inline stan/math/prim/scal/meta/partials_return_type.hpp]: Just to clarify the doc a little, this computes the type of the partial derivative. Perhaps: "base scalar return type" -> "partial derivative type"

Maybe there's a better way to state that, but I think that would a little more accurate. I don't know if that'll only make sense to me. Please feel free to ignore or do replace with something else.

File: test/unit/math/prim/scal/meta/return_type_test.cpp [hunk 8]
```diff
@@ -15,3 +20,64 @@ TEST(MetaTraits, ReturnTypeFloat) {
 TEST(MetaTraits, ReturnTypeInt) {
   test::expect_same_type<double, return_type<int>::type>();
 }
+
+TEST(MetaTraits, ReturnTypeVar) {
```
Comment (syclik) [inline test/unit/math/prim/scal/meta/return_type_test.cpp]: Ahh... so you've hit the confusing part of our source layout. The `var` tests belong in the `rev` folder. The `fvar<double>` tests belong in the `fwd` folder. The  `fvar<fvar<var>>` tests belong in the `mix` folder.

I can help. Mind giving me permissions on your branch and I'll push directly there?

File: stan/math/prim/scal/meta/partials_return_type.hpp (hunk 1 above)
Comment (dpsimpson) [reply stan/math/prim/scal/meta/partials_return_type.hpp]: Yep. I got it wrong but it's right in the current version. (Misleading result due to having the wrong set of `#include`s.

File: test/unit/math/prim/scal/meta/return_type_test.cpp (hunk 8 above)
Comment (dpsimpson) [reply test/unit/math/prim/scal/meta/return_type_test.cpp]: Easy. I see what I did wrong now. I'll fix all of that.

File: stan/math/prim/scal/meta/partials_return_type.hpp (hunk 7 above)
Comment (dpsimpson) [reply stan/math/prim/scal/meta/partials_return_type.hpp]: Fixed

File: stan/math/prim/scal/meta/partials_return_type.hpp [hunk 9]
```diff
@@ -2,21 +2,50 @@
 #define STAN_MATH_PRIM_SCAL_META_PARTIALS_RETURN_TYPE_HPP
 
 #include <stan/math/prim/scal/meta/partials_type.hpp>
+#include <stan/math/rev/core/var.hpp>
```
[the remark this answers is not in the document]
Comment (dpsimpson) [reply stan/math/prim/scal/meta/partials_return_type.hpp]: Thanks. All fixed.

**Comment by dpsimpson** [comment]
> For a stan::math::var, the partial should be double.
> 
> For a stan::math::fvar<double>, it should be double. (I think if it was fvar<float>, it should return double, but we don't need to check this at every level of nesting.)
> 
> For a stan::math::fvar<var>, isn't it var? Or am I missing something? The directional derivative returned is a var (so we can autodiff through that).

This is correct. It took me a few goes to notice the `fvar<T>` specialization, so the tests were wrong. And I didn't understand where to put them. Everything is good now. I've split all the tests and put them into the correct place.

I replaced the existing tests because these provide sufficient coverage. In the old `math/mix/scal/meta/partials_return_type_test.cpp` there were tests that checked with multiple types for autodiff variable. For example `partials_return_type<double, double, var, fvar<fvar<var> > >`. As @bob-carpenter said this can't happen, I removed the test.

**Comment by bob-carpenter** [comment]
That's tested all over the place, so it's going to be a pain to get rid of that extra is_nan check.  It's terrible for performance and I don't think it's the right behavior.  Hopefully we'll be able to fix that after the testing framework's in place.

**Comment by dpsimpson** [comment]
Copyright sorted. Tests run. Let me know if there’s anything else needed.

**Comment by bob-carpenter** [comment]
Thanks!  @syclik needs to sign off on the changes now.

File: stan/math/prim/scal/meta/partials_return_type.hpp [hunk 10]
```diff
@@ -2,21 +2,51 @@
 #define STAN_MATH_PRIM_SCAL_META_PARTIALS_RETURN_TYPE_HPP
 
 #include <stan/math/prim/scal/meta/partials_type.hpp>
+/*#include <stan/math/rev/core/var.hpp>
```
Comment (syclik) [inline stan/math/prim/scal/meta/partials_return_type.hpp]: Please remove commented lines. 

I can do this in a few mins.

**Comment by syclik** [comment]
@dpsimpson, thanks! And thanks for getting to everything before I had a chance to help you out. .
Right now, there's some commented includes. (According to what I'm seeing on my phone.) That should go, but it's easy. It's approved once that's done.

**Comment by syclik** [comment]
(of course, if you can change it before I do, then awesome!)

**Comment by syclik** [comment]
Just pushed. Then approved.

Thanks!

**Comment by dpsimpson** [comment]
Thanks! Sorry about those includes.

**Comment by syclik** [comment]
No prob! I'm at a conference and didn't get a chance to help earlier.

**Comment by syclik** [comment]
Thanks for the heads up. I'll try that directly next time.
